client/src/websockets.rs

```diff
@@ -0,0 +1,632 @@
+            }
+        });
+
+        //Add future to hashmap
+        self.outgoing_map
+            .lock()
+            .await
+            .insert(id, promise.abort_handle());
+
+        promise.await;
+
+        if let Some(incoming) = self.incoming_response.lock().await.get(&id) {
+            let status = match incoming.status {
+                Some(status) => status,
+                _ => panic!(),
+            };
+
+            if (200..300).contains(&status) {
+                Ok(incoming.clone())
+            } else {
+                anyhow::bail!("Got a bad response")
+            }
+        } else {
+            anyhow::bail!("Incoming response missing")
+        }
+    }
+
+    //TODO: Send TLS close_notify when closing connection
+    pub async fn close(&mut self, code: u16, reason: String) -> Result<()> {
+        Ok(self
+            .socket
+            .lock()
+            .await
+            .stream
+            .close(Some(frame::CloseFrame {
+                code: code.into(),
+                reason: reason.into(),
+            }))
+            .await?)
+    }
+}
+
+pub struct SendRequestOptions {
+    verb: String,
+    path: String,
+    body: Option<Vec<u8>>,
+    timeout: Option<u32>,
+    headers: Option<Vec<(String, String)>>,
+}
+
+#[derive(Clone)]
+pub struct KeepAliveOptions {
+    path: Option<String>,
+}
+
+#[derive(Clone)]
+struct KeepaliveHandler {
+    keepalive_options: KeepAliveOptions,
+    last_alive_at: Arc<Mutex<Instant>>,
+    timer_map: Arc<Mutex<HashMap<u32, JoinHandle<()>>>>,
+    websocket_resource: Arc<Mutex<WebsocketHandler>>,
+}
+
+impl KeepaliveHandler {
+    pub fn new(
+        keepalive_options: KeepAliveOptions,
+        websocket_resource: Arc<Mutex<WebsocketHandler>>,
+    ) -> Self {
+        KeepaliveHandler {
+            keepalive_options,
+            last_alive_at: Arc::new(Mutex::new(Instant::now())),
+            timer_map: Arc::new(Mutex::new(HashMap::new())),
+            websocket_resource,
+        }
+    }
+
+    pub async fn reset_keepalive(&mut self) {
+        *self.last_alive_at.clone().lock().await = Instant::now();
+        let mut self_clone = self.clone();
+        self.set_timeout(KEEPALIVE_INTERVAL_MS).await;
+    }
+
+    //KeepAlive contains two send methods to replicate the functionality of Signal-Desktop to the best of my ability
+    async fn keepalive_super_send(&mut self) -> Result<bool> {
+        let sent_at = Instant::now();
+
+        let request_options = SendRequestOptions {
+            verb: "GET".to_string(),
+            path: "/v1/keepalive".to_string(),
+            body: None,
+            headers: None,
+            timeout: Some(KEEPALIVE_TIMEOUT_MS),
+        };
+
+        let result = self
+            .websocket_resource
+            .lock()
+            .await
+            .send_request(request_options)
+            .await?;
+
+        if let Some(status) = result.status {
+            if !(200..300).contains(&status) {
```

Comment:
As above eller make `const` (eller noget)

Kan godt være det bliver for grimt idk